microhapulator/marker.py

```diff
@@ -0,0 +1,232 @@
+            loci[marker.locus].add(marker)
+        return loci.values()
+
+    @staticmethod
+    def parse_sequences_from_fasta(path):
+        with mhopen(path, "r") as fh:
+            sequences = SeqIO.to_dict(SeqIO.parse(fh, "fasta"))
+            sequences = {seqid: record.seq for seqid, record in sequences.items()}
+        return sequences
+
+    def __getitem__(self, key):
+        if key in self.loci:
+            return self.loci[key]
+        elif key in self.markers:
+            return self.markers[key]
+        else:
+            raise KeyError(key)
+
+    def sequence(self, key):
+        for locus in self.loci.values():
+            if locus.id == key:
+                return self.sequences[locus.id]
+            for marker in locus:
+                if marker.id == key:
+                    return self.sequences[locus.id]
+        raise KeyError(key)
+
+    def __iter__(self):
+        for locusid, locus in sorted(self.loci.items()):
+            for marker in locus:
+                yield locus, marker
+
+    def validate(self, refrids=None, symmetric=False):
+        seqnames = set(refrids) if refrids is not None else set(self.sequences.keys())
+        locusnames = set(self.loci.keys())
+        if len(locusnames) == 0:
+            warn("no marker definitions in microhap index")
+            return
+        missing = locusnames - seqnames
+        if len(missing) > 0:
+            missing = ", ".join(sorted(missing))
+            message = f"reference sequences missing for the following loci: {missing}"
+            raise ValueError(message)
+        if symmetric:
+            extra = seqnames - locusnames
+            if len(extra) > 0:
+                extra = ", ".join(sorted(extra))
+                message = f"reference sequences with no marker definition: {extra}"
+                raise ValueError(message)
+
+    @property
+    def has_chrom_offsets(self):
+        positives, negatives = 0, 0
+        for locus, marker in self:
+            if marker.offsets_chrom is not None:
+                for offset in marker.offsets_chrom:
+                    if pd.isna(offset):
+                        negatives += 1
+                        break
+                else:
+                    positives += 1
+            else:
+                negatives += 1
+        if positives > 0 and negatives > 0:
+            message = "index includes marker definitions both with and without complete chromosome offset lists"
+            warn(message)
+        return negatives == 0
+
+
+class Locus:
+    def __init__(self, ident=None, markers=None, sequence=None):
+        self.identifier = None
+        if ident:
+            self.identifier = Identifier(ident)
+        self.marker_definitions = list()
+        if markers is not None:
+            for marker in markers:
+                self.marker_definitions.add(marker)
+        self.sequence = sequence
```

Comment:
Supporting class: essentially a list of marker definitions associated with the same...you guessed it...locus.